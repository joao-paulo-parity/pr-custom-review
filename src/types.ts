import { RestEndpointMethodTypes } from "@octokit/plugin-rest-endpoint-methods"

export type CommitState =
  RestEndpointMethodTypes["repos"]["createCommitStatus"]["parameters"]["state"]
