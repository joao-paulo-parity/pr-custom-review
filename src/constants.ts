import {
  AndRule,
  BasicRule,
  CommitState,
  OrRule,
  RulesConfigurations,
} from "./types"

export const commitStateSuccess: CommitState = "success"
export const commitStateFailure: CommitState = "failure"

// Fields used for detecting the different kinds of rules
export const simpleRuleUniqueFields: Array<keyof BasicRule> = [
  "min_approvals",
  "teams",
  "users",
]
export const andRuleUniqueFields: Array<keyof AndRule> = ["all"]
export const orRuleUniqueFields: Array<keyof OrRule> = ["any"]

export const rulesConfigurations: RulesConfigurations = {
  basic: {
    kind: "BasicRule",
    uniqueFields: ["min_approvals", "teams", "users"],
    invalidFields: ["any", "all"],
  },
  and: {
    kind: "AndRule",
    uniqueFields: ["all"],
    invalidFields: ["min_approvals", "teams", "users", "any"],
  },
  or: {
    kind: "OrRule",
    uniqueFields: ["any"],
    invalidFields: ["min_approvals", "teams", "users", "all"],
  },
}
