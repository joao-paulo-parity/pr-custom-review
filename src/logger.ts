import os from "os"
import { inspect } from "util"

// GitHub by default logs N lines when the job is starting.
// N = 4 for pr-custom-review:
// 1. Run org/pr-custom-review@tag
// 2.   with:
// 3.     token: ***
// 4.     config-file: ./.github/pr-custom-review-config.yml
const githubLogsInitialLineCount = 4

export class Logger {
  lineCount: number = 0
  public relevantStartingLine: number = 1

  constructor() {}

  doLog = (...args: any[]) => {
    const message = args
      .map(function (arg) {
        return typeof arg === "string" ? arg : inspect(arg)
      })
      .join(" ")
    process.stdout.write(`${message}${os.EOL}`)
    this.lineCount += (message.match(/\n/g) || "").length + 1
  }

  public log = (...args: any[]) => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
    this.doLog(...args)
  }

  public error = (...args: any[]) => {
    // Uses escape codes for displaying the error line as red.
    // https://dustinpfister.github.io/2019/09/19/nodejs-ansi-escape-codes/
    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
    this.doLog("\n\u001b[31;1;4mERROR: ", ...args, "\u001b[39m")
  }

  public warning = (...args: any[]) => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
    this.doLog("WARNING: ", ...args)
  }

  public failure = (...args: any[]) => {
    this.markNextLineAsRelevantStartingLine(
      // This uses delta = 2 because the error logging adds an extra newline at
      // the start
      2,
    )
    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
    this.error(...args)
  }

  public markNextLineAsRelevantStartingLine = (delta = 1) => {
    this.relevantStartingLine =
      this.lineCount + githubLogsInitialLineCount + delta
  }
}

export default Logger
