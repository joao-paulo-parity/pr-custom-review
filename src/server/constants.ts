export const ciJobName = "pr-custom-review"
