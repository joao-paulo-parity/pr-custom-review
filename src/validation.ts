import Joi from "joi"

import { Configuration, Rule } from "./types"

const ruleSchema = Joi.object<Rule>().keys({
  name: Joi.string().required(),
  condition: Joi.string().required(),
  check_type: Joi.string().valid("diff", "changed_files").required(),
  min_approvals: Joi.number().required(),
  users: Joi.array().items(Joi.string()).optional().allow(null),
  teams: Joi.array().items(Joi.string()).optional().allow(null),
})

export const configurationSchema = Joi.object<Configuration>().keys({
  rules: Joi.array().items(ruleSchema).required(),
})
