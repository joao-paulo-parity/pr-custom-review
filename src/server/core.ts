import fetch from "node-fetch"

import { ciJobName } from "./constants"
import { ServerContext } from "./types"

export const triggerCiJob = async (
  { gitlab, logger }: ServerContext,
  pr: { number: number; base: { repo: { name: string } } },
) => {
  const projectApi = `https://${
    gitlab.domain
  }/api/v4/projects/${encodeURIComponent(
    `${gitlab.targetGroup}/${pr.base.repo.name}`,
  )}`

  const targetJobRef = pr.number.toString()

  logger.info(
    { projectApi, pr, jobRef: targetJobRef },
    `Attempting to find the latest "${ciJobName}" job to retrigger`,
  )

  const gitlabHeaders = { "PRIVATE-TOKEN": gitlab.accessToken }

  let jobsPage = 1
  while (true) {
    const responseBody = await (
      await fetch(`${projectApi}/jobs?page=${jobsPage}&per_page=100`, {
        headers: gitlabHeaders,
      })
    ).text()
    const jobs = JSON.parse(responseBody) as {
      id: number
      name: string
      ref: string
    }[]

    if (jobs.length === 0) {
      break
    }

    for (const job of jobs) {
      if (job.ref === targetJobRef && job.name === ciJobName) {
        await fetch(`${projectApi}/jobs/${job.id}/retry`, {
          method: "POST",
          headers: gitlabHeaders,
        })
        return
      }
    }

    jobsPage++
  }
}
