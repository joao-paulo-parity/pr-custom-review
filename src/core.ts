import core from "@actions/core"
import fs from "fs"
import YAML from "yaml"

import Logger from "./logger"
import { Octokit, PR, RuleUserInfo } from "./types"
import { configurationSchema } from "./validation"

const combineUsers = async function (
  pr: PR,
  client: Octokit,
  presetUsers: string[],
  teams: string[],
) {
  const users: Map<string, RuleUserInfo> = new Map()

  for (const user of presetUsers) {
    if (pr.user.login != user) {
      users.set(user, { team: null })
    }
  }

  for (const team of teams) {
    const teamMembersResponse = await client.rest.teams.listMembersInOrg({
      org: pr.base.repo.owner.login,
      team_slug: team,
    })
    if (teamMembersResponse.status !== 200) {
      return new Error(`Failed to fetch team members from ${team}`)
    }

    for (const member of teamMembersResponse.data) {
      if (member === null) {
        continue
      }
      if (
        pr.user.login != member.login &&
        // We do not want to register a team for this user if their approval is
        // supposed to be requested individually
        users.get(member.login) === undefined
      ) {
        users.set(member.login, { team })
      }
    }
  }

  return users
}

export const runChecks = async function (
  pr: PR,
  octokit: Octokit,
  logger: Logger,
): Promise<"failure" | "success"> {
  const diffResponse: { data: string; status: number } = await octokit.request(
    pr.diff_url,
  )
  if (diffResponse.status !== 200) {
    logger.failure(
      `Failed to get the diff from ${pr.diff_url} (code ${diffResponse.status})`,
    )
    logger.log(diffResponse.data)
    return "failure"
  }
  const { data: diff } = diffResponse

  const changedFilesResponse = await octokit.request(
    "GET /repos/{owner}/{repo}/pulls/{pull_number}/files",
    {
      owner: pr.base.repo.owner.login,
      repo: pr.base.repo.name,
      pull_number: pr.number,
    },
  )
  if (changedFilesResponse.status !== 200) {
    logger.failure(
      `Failed to get the changed files from ${pr.html_url} (code ${changedFilesResponse.status})`,
    )
    logger.log(changedFilesResponse.data)
    return "failure"
  }
  const { data: changedFilesData } = changedFilesResponse
  const changedFiles = new Set(changedFilesData.map(({ filename }) => filename))
  logger.log("Changed files", changedFiles)

  type MatchedRule = {
    name: string
    min_approvals: number
    users: Map<string, RuleUserInfo>
  }
  const matchedRules: MatchedRule[] = []

  // Built in condition to search files with changes to locked lines
  const lockExpression = /ðŸ”’.*(\n^[+|-])|^[+|-].*ðŸ”’/gm
  if (lockExpression.test(diff)) {
    logger.log("Diff has changes to ðŸ”’ lines or lines following ðŸ”’")
    const users = await combineUsers(pr, octokit, [], ["pr-custom-review-team"])
    if (users instanceof Error) {
      logger.failure(users)
      return "failure"
    }
    matchedRules.push({ name: "LOCKS TOUCHED", min_approvals: 2, users })
  }

  const configFilePath = core.getInput("config-file")
  if (configFilePath === null || configFilePath.length === 0) {
    logger.log("No config file provided")
  } else if (fs.existsSync(configFilePath)) {
    const configFile = fs.readFileSync(configFilePath, "utf8")

    const validation_result = configurationSchema.validate(
      YAML.parse(configFile),
    )
    if (validation_result.error) {
      logger.failure("Configuration file is invalid")
      logger.log(validation_result.error)
      return "failure"
    }
    const config = validation_result.value

    for (const rule of config.rules) {
      const condition: RegExp = new RegExp(rule.condition, "gm")

      let matched = false
      switch (rule.check_type) {
        case "changed_files": {
          changedFilesLoop: for (const file of changedFiles) {
            if (condition.test(file)) {
              logger.log(
                `Matched expression "${rule.condition}" of rule "${rule.name}" for the file ${file}`,
              )
              matched = true
              break changedFilesLoop
            }
          }
          break
        }
        case "diff": {
          if (condition.test(diff)) {
            logger.log(
              `Matched expression "${rule.condition}" of rule "${rule.name}" on diff`,
            )
            matched = true
          }
          break
        }
        default: {
          const exhaustivenessCheck: never = rule.check_type
          logger.failure(`Check type is not handled: ${exhaustivenessCheck}`)
          return "failure"
        }
      }
      if (!matched) {
        continue
      }

      const users = await combineUsers(
        pr,
        octokit,
        rule.users ?? [],
        rule.teams ?? [],
      )
      if (users instanceof Error) {
        logger.failure(users)
        return "failure"
      }
      matchedRules.push({
        name: rule.name,
        min_approvals: rule.min_approvals,
        users,
      })
    }
  } else {
    logger.failure(`Could not read config file at ${configFilePath}`)
    return "failure"
  }

  if (matchedRules.length !== 0) {
    const reviewsResponse = await octokit.rest.pulls.listReviews({
      owner: pr.base.repo.owner.login,
      repo: pr.base.repo.name,
      pull_number: pr.number,
    })
    if (reviewsResponse.status !== 200) {
      logger.failure(
        `Failed to fetch reviews from ${pr.html_url} (code ${reviewsResponse.status})`,
      )
      logger.log(reviewsResponse.data)
      return "failure"
    }
    const { data: reviews } = reviewsResponse

    const latestReviews: Map<
      number,
      { id: number; user: string; isApproval: boolean }
    > = new Map()
    for (const review of reviews) {
      if (review.user === null || review.user === undefined) {
        continue
      }
      const prevReview = latestReviews.get(review.user.id)
      if (
        prevReview === undefined ||
        // The latest review is the one with the highest id
        prevReview.id < review.id
      ) {
        latestReviews.set(review.user.id, {
          id: review.id,
          user: review.user.login,
          isApproval: review.state === "APPROVED",
        })
      }
    }
    logger.log("latestReviews", latestReviews.values())

    const problems: string[] = []

    const usersToAskForReview: Map<string, RuleUserInfo> = new Map()
    let highestMinApprovalsRule: MatchedRule | null = null
    for (const rule of matchedRules) {
      if (rule.users.size !== 0) {
        const approvedBy: Set<string> = new Set()
        for (const review of latestReviews.values()) {
          if (rule.users.has(review.user) && review.isApproval) {
            approvedBy.add(review.user)
          }
        }
        if (approvedBy.size < rule.min_approvals) {
          const missingApprovals: {
            username: string
            team: string | null
          }[] = []
          for (const [username, { team }] of rule.users) {
            if (!approvedBy.has(username)) {
              missingApprovals.push({ username, team })
              const prevUser = usersToAskForReview.get(username)
              if (
                // Avoid registering the same user twice
                prevUser === undefined ||
                // If the team is null, this user was not asked as part of a
                // team, but individually. They should always be registered with
                // "team: null" that case to be sure the review will be
                // requested individually, even if they were previously
                // registered as part of a team.
                team === null
              ) {
                usersToAskForReview.set(username, { team })
              }
            }
          }
          problems.push(
            `Rule "${rule.name}" needs at least ${
              rule.min_approvals
            } approvals, but ${
              approvedBy.size
            } were matched. The following users have not approved yet: ${missingApprovals
              .map(function (user) {
                return `${
                  user.username
                }${user.team ? ` (team: ${user.team})` : ""}`
              })
              .join(", ")}.`,
          )
        }
      } else if (
        highestMinApprovalsRule === null ||
        highestMinApprovalsRule.min_approvals < rule.min_approvals
      ) {
        highestMinApprovalsRule = rule
      }
    }

    if (usersToAskForReview.size !== 0) {
      logger.log("usersToAskForReview", usersToAskForReview)
      const teams: Set<string> = new Set()
      const users: Set<string> = new Set()
      for (const [user, { team }] of usersToAskForReview) {
        if (team === null) {
          users.add(user)
        } else {
          teams.add(team)
        }
      }
      await octokit.request(
        "POST /repos/{owner}/{repo}/pulls/{pull_number}/requested_reviewers",
        {
          owner: pr.base.repo.owner.login,
          repo: pr.base.repo.name,
          pull_number: pr.number,
          reviewers: Array.from(users),
          team_reviewers: Array.from(teams),
        },
      )
    }

    if (highestMinApprovalsRule !== null) {
      let approvalCount = 0
      for (const review of latestReviews.values()) {
        if (review.isApproval) {
          approvalCount++
        }
      }
      if (approvalCount < highestMinApprovalsRule.min_approvals) {
        problems.push(
          `Rule ${highestMinApprovalsRule.name} requires at least ${highestMinApprovalsRule.min_approvals} approvals, but only ${approvalCount} were given`,
        )
      }
    }

    if (problems.length !== 0) {
      logger.failure("The following problems were found:")
      for (const problem of problems) {
        logger.log(problem)
      }
      logger.log("")
      return "failure"
    }
  }

  return "success"
}
