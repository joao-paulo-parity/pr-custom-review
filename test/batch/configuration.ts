import { Octokit } from "@octokit/rest"
import nock from "nock"
import YAML from "yaml"

import {
  basePR,
  condition,
  configFileContentsApiPath,
  configFilePath,
  githubApi,
  githubWebsite,
  team,
} from "test/constants"
import Logger from "test/logger"

import { rulesConfigurations } from "src/constants"
import { runChecks } from "src/core"

describe("Configuration", function () {
  let logger: Logger
  let octokit: Octokit
  let logHistory: string[]

  beforeEach(function () {
    nock.disableNetConnect()
    logHistory = []
    logger = new Logger(logHistory)
    octokit = new Octokit()
    nock(githubWebsite)
      .get(basePR.diff_url.slice(githubWebsite.length))
      .reply(200, condition)
  })

  for (const [missingField, value] of [
    ["name", condition],
    ["condition", condition],
    ["check_type", "diff"],
    ["min_approvals", 1],
  ]) {
    it(`Configuration is invalid if ${missingField} is missing`, async function () {
      nock(githubApi)
        .get(configFileContentsApiPath)
        .reply(200, {
          content: Buffer.from(
            `
            rules:
              - ${missingField === "name" ? "" : `name: ${value}`}
                ${missingField === "condition" ? "" : `condition: ${value}`}
                ${missingField === "check_type" ? "" : `check_type: ${value}`}
                ${
                  missingField === "min_approvals"
                    ? ""
                    : `min_approvals: ${value}`
                }
            `,
          ).toString("base64"),
        })

      expect(
        await runChecks(basePR, octokit, logger, {
          configFilePath,
          locksReviewTeam: team,
        }),
      ).toBe("failure")

      expect(logHistory).toMatchSnapshot()
    })
  }

  it("Configuration is invalid if min_approvals is less than 1", async function () {
    nock(githubApi)
      .get(configFileContentsApiPath)
      .reply(200, {
        content: Buffer.from(
          `
          rules:
            - name: ${condition}
              condition: ${condition}
              check_type: diff
              min_approvals: 0
          `,
        ).toString("base64"),
      })

    expect(
      await runChecks(basePR, octokit, logger, {
        configFilePath,
        locksReviewTeam: team,
      }),
    ).toBe("failure")

    expect(logHistory).toMatchSnapshot()
  })

  it("Configuration is invalid if locksReviewTeam is empty or missing", async function () {
    expect(
      await runChecks(basePR, octokit, logger, {
        configFilePath,
        locksReviewTeam: "",
      }),
    ).toBe("failure")

    expect(logHistory).toMatchSnapshot()
  })

  for (const { kind, invalidFields } of Object.values(rulesConfigurations)) {
    const goodRule = (function () {
      switch (kind) {
        case "BasicRule": {
          return {
            name: condition,
            condition: condition,
            check_type: "diff",
            min_approvals: 1,
          }
        }
        case "AndRule": {
          return {
            name: condition,
            condition: condition,
            check_type: "diff",
            all: [],
          }
        }
        case "OrRule": {
          return {
            name: condition,
            condition: condition,
            check_type: "diff",
            any: [],
          }
        }
        default: {
          const exhaustivenessCheck: never = kind
          throw new Error(`Rule kind is not handled: ${exhaustivenessCheck}`)
        }
      }
    })()

    for (const invalidField of invalidFields) {
      const invalidFieldValidValue = (function () {
        switch (invalidField) {
          case "all":
          case "teams":
          case "users":
          case "any": {
            return []
          }
          case "min_approvals": {
            return 1
          }
          default: {
            const exhaustivenessCheck: never = invalidField
            throw new Error(
              `invalidField is not handled: ${exhaustivenessCheck}`,
            )
          }
        }
      })()

      it(`Rule kind ${kind} does not allow invalid field ${invalidField}`, async function () {
        const badRule = { ...goodRule, [invalidField]: invalidFieldValidValue }

        nock(githubApi)
          .get(configFileContentsApiPath)
          .reply(200, {
            content: Buffer.from(YAML.stringify({ rules: [badRule] })).toString(
              "base64",
            ),
          })

        expect(
          await runChecks(basePR, octokit, logger, {
            configFilePath,
            locksReviewTeam: team,
          }),
        ).toBe("failure")

        expect(logHistory).toMatchSnapshot()
      })
    }
  }

  it("Configuration is invalid if min_approvals is less than 1", async function () {
    nock(githubApi)
      .get(configFileContentsApiPath)
      .reply(200, {
        content: Buffer.from(
          `
          rules:
            - name: ${condition}
              condition: ${condition}
              check_type: diff
              min_approvals: 0
          `,
        ).toString("base64"),
      })

    expect(
      await runChecks(basePR, octokit, logger, {
        configFilePath,
        locksReviewTeam: team,
      }),
    ).toBe("failure")

    expect(logHistory).toMatchSnapshot()
  })

  afterEach(function () {
    nock.cleanAll()
    nock.enableNetConnect()
  })
})
