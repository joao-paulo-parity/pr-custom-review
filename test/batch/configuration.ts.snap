// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration Configuration is invalid if check_type is missing 1`] = `
Array [
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].check_type\\" is required] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].check_type\\" is required',
      path: [Array],
      type: 'any.required',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Configuration is invalid if condition is missing 1`] = `
Array [
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].condition\\" is required] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].condition\\" is required',
      path: [Array],
      type: 'any.required',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Configuration is invalid if locksReviewTeam is empty or missing 1`] = `
Array [
  "ERROR:  Locks Review Team should be provided",
]
`;

exports[`Configuration Configuration is invalid if min_approvals is less than 1 1`] = `
Array [
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].min_approvals\\" must be greater than or equal to 1] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].min_approvals\\" must be greater than or equal to 1',
      path: [Array],
      type: 'number.min',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Configuration is invalid if min_approvals is less than 1 2`] = `
Array [
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].min_approvals\\" must be greater than or equal to 1] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].min_approvals\\" must be greater than or equal to 1',
      path: [Array],
      type: 'number.min',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Configuration is invalid if min_approvals is missing 1`] = `
Array [
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].name\\" must be a string] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].name\\" must be a string',
      path: [Array],
      type: 'string.base',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Configuration is invalid if name is missing 1`] = `
Array [
  "ERROR:  Configuration file is invalid",
  "[Error [ValidationError]: \\"rules[0].name\\" is required] {
  _original: { rules: [ [Object] ] },
  details: [
    {
      message: '\\"rules[0].name\\" is required',
      path: [Array],
      type: 'any.required',
      context: [Object]
    }
  ]
}",
]
`;

exports[`Configuration Rule kind AndRule does not allow invalid field any 1`] = `
Array [
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"AndRule\\" because it had the field \\"all\\", but it also has the field \\"any\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind AndRule does not allow invalid field min_approvals 1`] = `
Array [
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"min_approvals\\", but it also has the field \\"all\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind AndRule does not allow invalid field teams 1`] = `
Array [
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"teams\\", but it also has the field \\"all\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind AndRule does not allow invalid field users 1`] = `
Array [
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"users\\", but it also has the field \\"all\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind BasicRule does not allow invalid field all 1`] = `
Array [
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"min_approvals\\", but it also has the field \\"all\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind BasicRule does not allow invalid field any 1`] = `
Array [
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"min_approvals\\", but it also has the field \\"any\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind OrRule does not allow invalid field all 1`] = `
Array [
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"AndRule\\" because it had the field \\"all\\", but it also has the field \\"any\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind OrRule does not allow invalid field min_approvals 1`] = `
Array [
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"min_approvals\\", but it also has the field \\"any\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind OrRule does not allow invalid field teams 1`] = `
Array [
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"teams\\", but it also has the field \\"any\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;

exports[`Configuration Rule kind OrRule does not allow invalid field users 1`] = `
Array [
  "ERROR:  Rule \\"condition\\" was expected to be of kind \\"BasicRule\\" because it had the field \\"users\\", but it also has the field \\"any\\", which belongs to another kind. Mixing fields from different kinds of rules is not allowed.",
]
`;
